### DISCUSSION TITLE
Are instruction memory and constant memory modeled int the functional model

### DISCUSSION THREAD

**Discussion by say-nothing** [accel-sim/accel-sim-framework#177] [Q&A]
Hi, All,

Hope you are doing well. I am trying to get the real program data from cuda-sim (the functional model) recently, and found that all the constant memory and instructions memory space can not be dump out from cuda-sim, did these two memory space modeled accurately in the cuda-sim? Or does the program instructions or constant memory been loaded into the functional model?

Thanks and best regards.
Shougang

**Comment by JRPan** [comment]
Not 100% sure, but I think these are not modeled in the functional model. You don't need to model the instruction memory space in the functional model. It's just loading instructions. There is no functional simulation of that. It's just simply a fetch of the instruction. And this also applies to the constant space. It's just loading constants. 

It is possible to print the instruction being simulated. But I'm not sure what you are trying to do with "real program data" of instruction memory. 

Thanks,
Junrui

**Reply by say-nothing** [reply]
Hi, Junrui,

I am working on memory compression recently, and I want to load all the real program data from functional model to calculate the compression ratio. That's why I am also trying to get the instruction memory and constant memory data. Basically I want to know for a workloads, how much memory space I can save by compressing all the data, including instructions and constant memory. I did get most program data from functional model, however, I did not see anything regarding instruction/constant memory in cuda-sim

Thanks and best regards.
Shougang

**Reply by JRPan** [reply]
Okay. For instruction and constant, the memory usage should be static (maybe?). For example, I think you can add something to calculate the memory footprint based on how many PTX instructions there are. Or you can parse the PTX file directly. 
Anyway, I don't think there is any existing stuff you can use. You would probably need to implement something yourself. 
But again, I could be wrong. You can ask the gpgpu-sim group to see if they have any suggestions. 

Thanks.

**Reply by say-nothing** [reply]
Thanks for your clarification, I will try. And will get back to the community if I find anything new.
